src/main/java/org/truffleruby/core/kernel/KernelNodes.java

```diff
@@ -664,6 +657,32 @@ protected boolean assignsNoNewVariables(RootNodeWrapper rootNode) {
         private boolean frameHasOnlySelf(final FrameDescriptor descriptor) {
             return descriptor.getSize() == 1 && SelfNode.SELF_IDENTIFIER.equals(descriptor.getSlots().get(0).getIdentifier());
         }
+
+        @TruffleBoundary
+        public static String offsetSource(String method, String source, String file, int line) {
```

Comment:
Looks to me like this should return a Source, any reason not to?

Reply:
I try to make each method know about as few classes as possible to reduce complexity. The caller of this method needs to create a `Source`, but this method can be simpler and only deal with the strings, not knowing anything about sources.

Reply:
But then each caller duplicates the creation of a Source, and this method seems to already have all the information needed to create a Source, so I think it's better to create the Source here.
Furthermore, maybe we could have an abstraction on top of Source to deal with these weird line numbers, or at some point integrate such a change in Source itself, and then it would still make sense to keep this method to deal with negative line numbers, etc.
It could be 2 different methods (one private for String processing and one building the Source from the information), of course.

Reply:
> so I think it's better to create the Source here

Well I'll create another help method to create the `Source` that they can all use then.

> Furthermore, maybe we could have an abstraction on top of Source to deal with these weird line numbers, or at some point integrate such a change in Source itself

Yes there's an issue open for offset line numbers, including negative, GR-9504. I don't want to add our own abstraction around source if we can avoid it, and I don't want to add it temporarily.